lib/ansible/module_utils/network/checkpoint/checkpoint.py

```diff
@@ -57,3 +60,125 @@ def install_policy(connection, policy_package, targets):
+    for key, value in payload.items():
+        if value and isinstance(api_call_object.get(key), str) and value != api_call_object.get(key):
+            return True
+    return False
+
+
+def run_api_command(connection, command, payload):
+    code, response = connection.send_request('/web_api/' + command, payload)
+
+    return code, response
+
+
+def get_payload_from_user_parameters(module, user_parameters):
+    payload = {}
+    for parameter in user_parameters:
+        if module.params[parameter]:
+            payload[parameter.replace("_", "-")] = module.params[parameter]
+    return payload
+
+
+def api_command(module, command, user_parameters):
+    payload = get_payload_from_user_parameters(module, user_parameters)
+    connection = Connection(module._socket_path)
+    code, response = run_api_command(connection, command, payload)
+    result = {'changed': True}
+
+    if code == 200:
+        result['checkpoint_' + command.replace("-", "_")] = response
+    else:
+        module.fail_json(msg='Checkpoint device returned error {0} with message {1}'.format(code, response))
+
+    module.exit_json(**result)
+
+
+def api_call_facts(module, api_call_object, user_parameters):
+    payload = get_payload_from_user_parameters(module, user_parameters)
+    file_name_plural = "checkpoint_" + api_call_object.replace("_", "-") + "s"
+    if payload.get("name") is None and payload.get("uid") is None:
+        api_call_object += "s"
+    connection = Connection(module._socket_path)
+    code, response = get_api_call_object(connection, api_call_object, payload)
+    if code == 200:
+        module.exit_json(ansible_facts={file_name_plural: response})
+    else:
+        module.fail_json(msg='Checkpoint device returned error {0} with message {1}'.format(code, response))
+
+
+def api_call(module, api_call_object, user_parameters, unique_payload_for_get):
+    payload = get_payload_from_user_parameters(module, user_parameters)
+    file_name_plural = "checkpoint_" + api_call_object.replace("_", "-") + "s"
+    connection = Connection(module._socket_path)
+    code, response = get_api_call_object(connection, api_call_object, unique_payload_for_get)
+    result = {'changed': False}
+
+    if module.params['state'] == 'present':
+        if code == 200:
+            if needs_update(payload, response):
+                code, response = update_api_call_object(connection, api_call_object, payload)
+                if module.params['auto_publish_session']:
+                    publish(connection)
+
+                    if module.params['auto_install_policy']:
+                        install_policy(connection, module.params['policy_package'], module.params['targets'])
+
+                result['changed'] = True
+                result[file_name_plural] = response
+            else:
+                pass
+        elif code == 404:
+            code, response = create_api_call_object(connection, api_call_object, payload)
+
```

Comment:
please handle the exception response, if the response code is not `200` then you can throw the error response like: `self.module.fail_json(msg=response))` for all the api_calls.

Reply:
Hi, that's not an error, because I get to this code if the user want to add object (state='present'), and the object is not on the database of checkpoint, and that's why the code is 404, so we add the new object..

Reply:
@chkp-orso I understand that, but the thing is if there is exception response from function `create_api_call_object` you are not handling that response, and this is why `checkpoint_host` module was not throwing error response when u tested with `invalid ip_addr`, so this was the fix that I had added in PR.

Reply:
ohh I see now what you mean, but the response itself which I get from "code, response = create_api_call_object(connection, api_call_object, payload)"
is still not the response of checkpoint, just "'Object not found'"

Reply:
@chkp-orso I have verified the changes in the PR and the response exception in case of incorrect `ip_address`  was same as thrown from `checkpoint` api, also if u don't implement the respective check then in that case u'll receive `Object Not found` error from here: [AnsibleConnectionFailure](https://github.com/ansible/ansible/blob/97ac03f613bf747c07e526f012044323077162f1/lib/ansible/plugins/httpapi/checkpoint.py#L66)

Reply:
@justjais so how can I get this response? I tried:

code, response = add_api_call_object(connection, api_call_object, payload)
if code is 404:
     module.fail_json(msg=response)

and I also tried:

            try:
                code, response = add_api_call_object(connection, api_call_object, payload)
            except Exception as e:
                module.fail_json(msg=response)

but I don't get the checkpoint response

Reply:
@chkp-orso u are not able to get the error response from CP, coz the response code is not `404` when incorrect IP address is given as input its `400`, change the if check as in the PR:
```
if code != 200:
  module.fail_json(msg=response)
```
and respective check shall handle all the error code other than success i.e. 200 and shall throw the respective CP error

Reply:
@justjais ohh I see, I changed it but now I now I get only part of checkpoint output.
The output is a dictionary, and I get only the message key, I need somehow to get the whole dictionary output
Thanks!

Reply:
@chkp-orso, can you please install python `q` module and debug the `code n response` that u are getting once you are calling the CP api. The way you can debug the code is once u install the python `q` module u can import the `q` module in the code where you want to debug, then u can directly put debug statements via `q` module as:
`q("Debug Statement!")` and `q(code, response)`
and this should generate debug values under the TMPDIR plz refer following link for further info [q module](https://pypi.org/project/q/).
Using this u can detect what exactly is going wrong.

Reply:
Thanks I didn't know this module. 
So debug and this is the values of code and response:

      code=404,
      response=u'Could not connect to https://172.23.78.74:443/web_api/add-network: Bad Request'

and it's only the first part of the error request of checkpoint :(

Reply:
@chkp-orso can u run the same scenario which you are trying via some restclient and copy-paste the output here for me to understand the exact issue here.

Reply:
@justjais If I run the same scenario (when subnet is illegal) threw postman (web-api) the output is the following json:
`{
    "code": "err_validation_failed",
    "message": "Validation failed with 1 warning",
    "warnings": [
        {
            "message": "The IPv4 address is invalid based on the defined net mask. The suggested address: 2.2.2.0 will be used instead."
        }
    ]
}`

Reply:
@chkp-orso plz debug the `httpapi` connection plugin via `q` module as why there is difference in the response captured from `send_request`.

Reply:
I tried it, but it's not recognize the q module.
The following error is appear: 

File "/home/admin/ansible/lib/ansible/plugins/httpapi/checkpoint.py", line 57, in send_request
    import q; q("send request")
ImportError: No module named q

Reply:
@chkp-orso it seems to be some python/Ansible environment issue, as I am not facing the respective issue and I am able to debug the respective lines. Can you please try to triage the respective issue.

Reply:
@justjais  I tried for a few days but didn't succeeded so far :(

Reply:
maybe the cause of this problem is the rpc method, means `send_request` function is called from plugins/httpapi/__init__.py via the Connection object **rpc** method as per the ansible design